### DISCUSSION TITLE
MAUI Staggered CollectionView

### DISCUSSION THREAD

**Discussion by Elinares-82** [CommunityToolkit/Maui#1764] [General]
There are a couple of request to have in MAUI event in XF the ability to create a Staggered CollectionView. I saw this in several mobile frameworks, however the Micrsoft team is not pretending to add this feature to MAUI. These are the two request feature links that ask for this. Is there an intention to add Staggered layout to the CollectionView?

https://github.com/dotnet/maui/issues/7761
https://github.com/dotnet/maui/issues/9659

**Comment by VladislavAntonyuk** [comment]
We need to implement a new layout for CollectionView in addition to Linear/GridItemsLayout.

I like the idea.
For now you can use: https://github.com/jsuarezruiz/netmaui-beautyshop-app-challenge

**Reply by Elinares-82** [reply]
Thanks ! However I have read that this control has some bugs and it’s not ready for large amount of data . I have an e-commerce platform in XF using FlowLisView and I moved to Maui using CollectionView, having said this , can I use that control for production purposes ?

**Reply by bijington** [reply]
> Thanks ! However I have read that this control has some bugs and it’s not ready for large amount of data . I have an e-commerce platform in XF using FlowLisView and I moved to Maui using CollectionView, having said this , can I use that control for production purposes ?

I think that will be up to you to evaluate whether it covers your business needs and is stable for it.

**Reply by Elinares-82** [reply]
Thanks ! Sure, I’ll give it a try. I hope the MAUI team does some effort with this kind of control .

**Comment by GeorgeLeithead** [comment]
I think I raised this before under #572

**Reply by Elinares-82** [reply]
Yes @GeorgeLeithead , but the grid control called VariableSizedWrapGrid is ready to production purposes? I mean in this comment https://github.com/dotnet/maui/issues/7761#issuecomment-1158789389   jsuarezruiz said that _**"if the number of elements is not very large"**_ but most of the application can get a lot of items for example an ecommerce platform (my case). Is there an alternative for this type of layout?

**Reply by Elinares-82** [reply]
Is there news regarding this requirement?

**Reply by bijington** [reply]
You might have to try using that control and see how it performs for you

**Reply by Elinares-82** [reply]
Thanks, I'll give it a try.

**Reply by Elinares-82** [reply]
Hi, with the help of @roubachof 

I realized that there is a control that can help to create a great one. I'm working porting to MAUI.

[WCollectionView](https://github.com/leoholcman/WCollectionView.Forms)

**Comment by VladislavAntonyuk** [comment]
https://vladislavantonyuk.github.io/articles/.NET-MAUI-CollectionView-with-Staggered-Layout use as the start point. For iOS and Windows I will update a bit later

**Reply by Elinares-82** [reply]
I'll review it. But the is there android as well?

**Reply by VladislavAntonyuk** [reply]
I implemented only for Android and iOS. For Windows there is a CommunityToolkit.WinUI StaggeredLayout but it is not compatible with .NET MAUI ListViewBase.
I will update sample project with final solution.

**Reply by Elinares-82** [reply]
Hi @VladislavAntonyuk , I reviewed the Layout, and first thanks for your help, effort and dedication to improve the MAUI world. Now I'm trying to use your Layout but the challenge that I'm facing is, I have a control that uses the CollectionView but this control is in an separate project in Net8  called NeighborStore.Mobile.Controls, so the question is "How can I use the layout in a separate project since we cannot use the preprocessor directives like #if ANDROID"

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/32d8ff54-1a90-48f8-a38f-fd632833e364)

Thanks

**Reply by VladislavAntonyuk** [reply]
Hello, the code is all about handlers, so you can put handler in Platforms folder in your startup project. I use “if” directive to have only one file as the handler here is pretty simple. But anyway in your case you split controls and handlers in different projects

**Reply by Elinares-82** [reply]
Thanks!, That I'm doing first is moving from net8 library project to Net MAUI Library project, so maybe this will allow me to use the Handler easily.

**Reply by Elinares-82** [reply]
Another thing , I do not know if you already check this repo

[https://github.com/leoholcman/WCollectionView.Forms](url)

This can be helpful for the iOS part.

**Reply by VladislavAntonyuk** [reply]
Yes, I implemented android, ios, maccatalyst. Only windows left.

**Reply by Elinares-82** [reply]
Yes, I'm doing all the changes in my solution to test and use your Layout. however, in the ios side apparently there is a layout that have to be implemented. So that I'm assuming is that the StagLayout have to be create inherited from ItemsViewLayout

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/d1b8d299-f499-4dce-b03d-e58bbcf166bb)

**Reply by VladislavAntonyuk** [reply]
Yes, see the code in repository

**Reply by Elinares-82** [reply]
Thanks!, Reviewing and testing.

**Reply by Elinares-82** [reply]
Hi @VladislavAntonyuk  I'm testing the control, and so far so good in Android. The only thing I'm facing is with the EmptyView, apparently the empty view is not allowing to the ItemTemplate render correctly and all the items are displayed in single columns instead two. Also with the header template, The header is being displayed as an item rather a header.

This is the examples:

The header is being displayed as an item instead of wide slide I mean using the entire width.

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/1b5d46e6-a394-4a5a-8d19-d6a5acacdb19)


Should to be rendered as the following (this does not use the StaggeredLayout)

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/cbc0cacc-081b-4a2b-bb27-52307a29afe6)

And this is using the EmptyView

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/aa8ae4ec-9276-4e66-9335-3eb9fda917b7)

**Reply by Elinares-82** [reply]
@VladislavAntonyuk  did you have a chance to review this? I'm trying to figure out, but I do not know how to prevent the Layout rendering in the CollectionView header.

**Reply by VladislavAntonyuk** [reply]
I didn't check it, as I created it for a start point / demo purposes and not as a final solution. As @jfversluis  mentioned there will be some changes in the Collection View implementation. 

The workaround for now is extracting a header content out of the collection view.
Probably header is a part of view items in .NET MAUI implementation. You need to see how the GridLayout works.

**Reply by Elinares-82** [reply]
Thanks @VladislavAntonyuk , I was thinking to take the header out but I need to hide this when the collection is scrolling down. So, I'll continue to review this topic. And in fact I'm reviewing the GridLayout to find out how it works.

**Reply by Elinares-82** [reply]
Just a quick update. I found why the grid layout work just fine with the header and footer

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/349f1ae6-7c47-43fb-8f17-e2d1e2098a5f)

Now, the property to use with the StaggeredLayour is FullSpan when the item is Header or Footer as the Android docs says

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/b3a0f690-7abb-412f-bdc9-e39e89e705a5)

So I created my own adapter and override OnBindViewHolder method, I have almost everything I need, however, I was not able to  find the StaggeredGridLayoutManager.LayoutParams since always the LayoutParameters property is null

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/7f817d53-e2b5-403b-b66a-c5c004fe558c)

**Reply by VladislavAntonyuk** [reply]
Good found, I will take a look and update sample/article.

**Reply by Elinares-82** [reply]
Sure!, let me shared with you the cs

[StaggeredItemsLayout.zip](https://github.com/CommunityToolkit/Maui/files/15355173/StaggeredItemsLayout.zip)

**Reply by Elinares-82** [reply]
There are more functionalities that are not working for example RemainingItemsThresholdReachedCommand, but step by step I'll help to adding them.

**Reply by Elinares-82** [reply]
@VladislavAntonyuk  Quick update. I achieved the desired result by setting the layout.

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/92e26900-a63b-442f-bd86-3b5d8ab42a05)

A lot of work needs to be done on this, but I think this is the way.

**Reply by Elinares-82** [reply]
I added Layout type validation.

![image](https://github.com/CommunityToolkit/Maui/assets/158092565/a3c7f452-9964-47ad-9135-25139787db47)

And this is the file with all the changes.

[StaggeredItemsLayout.zip](https://github.com/CommunityToolkit/Maui/files/15359256/StaggeredItemsLayout.zip)

**Reply by Elinares-82** [reply]
Hi @VladislavAntonyuk  I had to stop working on this for a few weeks, but I would like to start again. Did you had the chance to review the code that I provided?

Regards,

**Reply by VladislavAntonyuk** [reply]
I have not seen your code yet. Enjoy your free time.
You can open pull request to my MauiSamples repo. We can merge new functionality and fixes there.
